File: src/DashSpec.Core/Parsing/Lexing/TokenReader.cs
using DashSpec.Core.Model;

namespace DashSpec.Core.Parsing;

internal sealed class TokenReader
{
    private readonly IReadOnlyList<Token> _tokens;
    private readonly Stack<BlockCloseStyle> _blockCloseStyles = new();
    private int _index;

    public TokenReader(IReadOnlyList<Token> tokens) => _tokens = tokens;

    internal void PushBlockClose(BlockCloseStyle style) => _blockCloseStyles.Push(style);

    internal void PopBlockClose()
    {
        if (_blockCloseStyles.Count > 0)
        {
            _blockCloseStyles.Pop();
        }
    }

    internal BlockCloseStyle PeekBlockClose() =>
        _blockCloseStyles.Count > 0 ? _blockCloseStyles.Peek() : BlockCloseStyle.EndKeyword;

    public string? ConsumedRuntimePath { get; private set; }

    /// <summary>Deprecated alias; use <see cref="ConsumedRuntimePath"/>.</summary>
    public string? ConsumedConfigPath => ConsumedRuntimePath;

    public SqlDialect ConsumedSqlDialect { get; private set; } = SqlDialect.TSql;

    public string? ConsumedDiagramLibraryPath { get; private set; }

    public string? ConsumedPalettePath { get; private set; }

    public void SkipFileDirectives()
    {
        SkipNewlines();
        while (IsAt(TokenKind.At))
        {
            Advance();
            if (TryKeyword("runtime") || TryKeyword("config"))
            {
                if (ConsumedRuntimePath is not null)
                {
                    throw new DashSpecParseException(
                        "Only one @runtime directive is allowed per .dashspec file.");
                }

                ConsumedRuntimePath = ReadString();
                SkipNewlines();
                continue;
            }

            if (TryKeyword("sqldialect"))
            {
                ConsumedSqlDialect = SqlDialectParser.Parse(ReadIdent());
                SkipNewlines();
                continue;
            }

            if (TryKeyword("diagramlibrary"))
            {
                if (ConsumedDiagramLibraryPath is not null)
                {
                    throw new DashSpecParseException(
                        "Only one @diagramlibrary directive is allowed per .dashspec file.");
                }

                ConsumedDiagramLibraryPath = ReadString();
                SkipNewlines();
                continue;
            }

            if (TryKeyword("palette"))
            {
                SkipNewlines();
                if (Current.Kind is TokenKind.String)
                {
                    if (ConsumedPalettePath is not null)
                    {
                        throw new DashSpecParseException(
                            "Only one @palette file directive is allowed per .dashspec file.");
                    }

                    ConsumedPalettePath = ReadString();
                    SkipNewlines();
                    continue;
                }

                _index -= 2;
                break;
            }

            _index--;
            break;
        }
    }

    public string ReadPropertyKey(bool allowQuoted = false)
    {
        SkipNewlines();
        if (allowQuoted && Current.Kind is TokenKind.String)
        {
            return ReadString();
        }

        return ReadIdent();
    }

    public bool IsEof => Current.Kind is TokenKind.Eof;

    public TokenKind CurrentKind
    {
        get
        {
            SkipNewlines();
            return Current.Kind;
        }
    }

    private Token Current => _tokens[_index];

    /// <summary>Current token kind without skipping newlines (use for same-line inline grammar).</summary>
    public TokenKind RawKind => Current.Kind;

    public bool IsOnNewline() => Current.Kind is TokenKind.Newline;

    public void SkipNewlines()
    {
        while (Current.Kind is TokenKind.Newline)
        {
            _index++;
        }
    }

    public void Advance() => _index++;

    public bool IsAt(TokenKind kind)
    {
        SkipNewlines();
        return Current.Kind == kind;
    }

    public void Expect(TokenKind kind)
    {
        SkipNewlines();
        if (Current.Kind != kind)
        {
            throw Unexpected(kind.ToString());
        }

        _index++;
    }

    public bool TryKeyword(string keyword)
    {
        SkipNewlines();
        if (Current.Kind is not TokenKind.Ident ||
            !string.Equals(Current.Value, keyword, StringComparison.OrdinalIgnoreCase))
        {
            return false;
        }

        _index++;
        return true;
    }

    public bool TryPeekIdent(out string ident)
    {
        SkipNewlines();
        if (Current.Kind is not TokenKind.Ident)
        {
            ident = string.Empty;
            return false;
        }

        ident = Current.Value;
        return true;
    }

    public readonly struct Mark
    {
        internal Mark(int index) => Index = index;
        internal int Index { get; }
    }

    public Mark CreateMark() => new(_index);

    public void Rewind(Mark mark) => _index = mark.Index;

    public bool TryModuleInclude(out string reference)
    {
        SkipNewlines();
        if (!IsAt(TokenKind.Bang))
        {
            reference = string.Empty;
            return false;
        }

        Advance();
        if (!TryKeyword("include"))
        {
            throw new DashSpecParseException("Expected 'include' after '!'.");
        }

        reference = ReadString();
        return true;
    }

    /// <summary>
    /// Optional postfix keyword on the current line only (does not cross <see cref="TokenKind.Newline"/>).
    /// Use for <c>ref</c>, optional trailing <c>as</c>, etc. — never for statement-start keywords.
    /// </summary>
    public bool TryKeywordSameLine(string keyword)
    {
        if (Current.Kind is TokenKind.Newline or TokenKind.Eof)
        {
            return false;
        }

        if (Current.Kind is not TokenKind.Ident ||
            !string.Equals(Current.Value, keyword, StringComparison.OrdinalIgnoreCase))
        {
            return false;
        }

        _index++;
        return true;
    }

    public string ReadIdentSameLine()
    {
        if (Current.Kind is not TokenKind.Ident)
        {
            throw Unexpected("identifier on the same line");
        }

        var value = Current.Value;
        _index++;
        return value;
    }

    public void ExpectKeyword(string keyword)
    {
        if (!TryKeyword(keyword))
        {
            throw Unexpected(keyword);
        }
    }

    public string ReadIdent()
    {
        SkipNewlines();
        if (Current.Kind is not TokenKind.Ident)
        {
            throw Unexpected("identifier");
        }

        var value = Current.Value;
        _index++;
        return value;
    }

    public string ReadQualifiedName()
    {
        SkipNewlines();
        var value = Current.Kind switch
        {
            TokenKind.Ident => Current.Value,
            TokenKind.Raw => Current.Value,
            _ => throw Unexpected("qualified name"),
        };

        _index++;
        return value;
    }

    public string ReadString()
    {
        SkipNewlines();
        if (Current.Kind is not TokenKind.String)
        {
            throw Unexpected("string");
        }

        var value = Current.Value;
        _index++;
        return value;
    }

    public string ReadScalarValue()
    {
        SkipNewlines();
        return Current.Kind switch
        {
            TokenKind.String => ReadString(),
            TokenKind.Ident => ReadIdent(),
            TokenKind.RelativeDay => ReadRelativeDay(),
            _ => throw Unexpected("scalar value"),
        };
    }

    public string ReadRawBlock()
    {
        SkipNewlines();
        if (Current.Kind is not TokenKind.Raw)
        {
            throw Unexpected("[[ … ]] raw block");
        }

        var value = Current.Value;
        _index++;
        return value;
    }

    public ColumnBindingValue ReadColumnBinding()
    {
        var column = ReadQualifiedName();
        if (TryKeyword("as"))
        {
            return new ColumnBindingValue(column, ReadString());
        }

        if (RawKind is TokenKind.Ident && !IsOnNewline())
        {
            var saved = SavePosition();
            var alias = ReadIdent();
            if (RawKind is TokenKind.Eq)
            {
                RestorePosition(saved);
                return new ColumnBindingValue(column, null);
            }

            if (string.Equals(alias, "end", StringComparison.OrdinalIgnoreCase))
            {
                RestorePosition(saved);
                return new ColumnBindingValue(column, null);
            }

            return new ColumnBindingValue(column, alias);
        }

        return new ColumnBindingValue(column, null);
    }

    public string ReadCommaSeparatedValues()
    {
        var parts = new List<string> { ReadListItem() };
        while (CurrentKind is TokenKind.Comma)
        {
            _index++;
            parts.Add(ReadListItem());
        }

        return string.Join(", ", parts);
    }

    public IReadOnlyList<string> ReadCommaListInline()
    {
        var names = new List<string> { ReadIdent() };
        while (CurrentKind is TokenKind.Comma)
        {
            _index++;
            names.Add(ReadIdent());
        }

        return names;
    }

    public string ReadDateDefaultValue()
    {
        var from = ReadDateBoundToken();
        if (RawKind is not TokenKind.DotDot)
        {
            return from;
        }

        _index++;
        var to = ReadDateBoundToken();
        return $"{from}..{to}";
    }

    private string ReadDateBoundToken()
    {
        SkipNewlines();
        return Current.Kind switch
        {
            TokenKind.RelativeDay => ReadRelativeDay(),
            TokenKind.Ident => ReadIdent(),
            _ => throw Unexpected("date bound (today, -Nd, yyyy-MM-dd)"),
        };
    }

    private string ReadRelativeDay()
    {
        SkipNewlines();
        if (Current.Kind is not TokenKind.RelativeDay)
        {
            throw Unexpected("relative day");
        }

        var value = Current.Value;
        _index++;
        return value;
    }

    private string ReadListItem()
    {
        SkipNewlines();
        return Current.Kind switch
        {
            TokenKind.Ident => ReadIdent(),
            TokenKind.String => ReadString(),
            _ => throw Unexpected("list item"),
        };
    }

    /// <summary>Reads scalar tokens until newline/EOF — does not cross line boundaries.</summary>
    public string ReadRestOfLine()
    {
        var parts = new List<string>();
        while (Current.Kind is TokenKind.Ident or TokenKind.Raw)
        {
            parts.Add(Current.Value);
            _index++;
        }

        return string.Join(' ', parts);
    }

    public DashSpecParseException Unexpected(string? expected = null)
    {
        var token = Current;
        var message = expected is null
            ? $"Unexpected token '{token.Value}' ({token.Kind}) at position {token.Start}."
            : $"Expected {expected}, got '{token.Value}' ({token.Kind}) at position {token.Start}.";
        return new DashSpecParseException(message, token.Start);
    }

    public int SavePosition() => _index;

    public void RestorePosition(int position) => _index = position;
}

File: src/DashSpec.Modeling.Parse/Lexing/TokenReader.fs
namespace DashSpec.Modeling.Parse.Lexing

open System
open System.Collections.Generic
open DashSpec.Modeling.Core

/// Token cursor over lexed dashspec input.
type TokenReader(tokens: IReadOnlyList<Token>) =
    let mutable index = 0
    let blockCloseStyles = Stack<BlockCloseStyle>()
    let mutable consumedRuntimePath: string option = None
    let mutable consumedDiagramLibraryPath: string option = None
    let mutable consumedPalettePath: string option = None

    member _.ConsumedRuntimePath = consumedRuntimePath
    member _.ConsumedDiagramLibraryPath = consumedDiagramLibraryPath
    member _.ConsumedPalettePath = consumedPalettePath

    member _.Current = tokens.[index]

    member _.RawKind = tokens.[index].Kind

    member _.IsOnNewline() = tokens.[index].Kind = TokenKind.Newline

    member private _.IsTrivia kind =
        kind = TokenKind.Newline || kind = TokenKind.LineComment || kind = TokenKind.BlockComment

    member this.SkipNewlines() =
        while this.IsTrivia tokens.[index].Kind do
            index <- index + 1

    member this.Advance() = index <- index + 1

    member this.IsEof =
        this.SkipNewlines()
        tokens.[index].Kind = TokenKind.Eof

    member this.IsAt kind =
        this.SkipNewlines()
        tokens.[index].Kind = kind

    member this.Expect kind =
        this.SkipNewlines()
        if tokens.[index].Kind <> kind then
            raise (this.Unexpected(string kind))
        index <- index + 1

    member this.TryKeyword keyword =
        this.SkipNewlines()
        if tokens.[index].Kind <> TokenKind.Ident then false
        elif not (String.Equals(tokens.[index].Value, keyword, StringComparison.OrdinalIgnoreCase)) then false
        else
            index <- index + 1
            true

    member this.TryKeywordSameLine keyword =
        match tokens.[index].Kind with
        | TokenKind.Newline | TokenKind.Eof -> false
        | TokenKind.Ident when String.Equals(tokens.[index].Value, keyword, StringComparison.OrdinalIgnoreCase) ->
            index <- index + 1
            true
        | _ -> false

    member this.ExpectKeyword keyword =
        if not (this.TryKeyword keyword) then
            raise (this.Unexpected keyword)

    member this.ReadIdent() =
        this.SkipNewlines()
        if tokens.[index].Kind <> TokenKind.Ident then
            raise (this.Unexpected "identifier")
        let value = tokens.[index].Value
        index <- index + 1
        value

    member this.ReadIdentSameLine() =
        if tokens.[index].Kind <> TokenKind.Ident then
            raise (this.Unexpected "identifier on the same line")
        let value = tokens.[index].Value
        index <- index + 1
        value

    member _.PushBlockClose style = blockCloseStyles.Push style

    member _.PopBlockClose() =
        if blockCloseStyles.Count > 0 then blockCloseStyles.Pop() |> ignore

    member _.PeekBlockClose() =
        if blockCloseStyles.Count > 0 then blockCloseStyles.Peek() else BlockCloseStyle.EndKeyword

    member _.SavePosition() = index

    member _.RestorePosition position = index <- position

    member this.Unexpected(?expected: string) =
        let token = tokens.[index]
        let message =
            match expected with
            | Some e -> $"Expected {e}, got '{token.Value}' ({token.Kind}) at position {token.Start}."
            | None -> $"Unexpected token '{token.Value}' ({token.Kind}) at position {token.Start}."
        DashSpec.Modeling.Core.DashSpecParseException(message, token.Start)

    member this.ReadString() =
        this.SkipNewlines()
        if tokens.[index].Kind <> TokenKind.String then
            raise (this.Unexpected "string")
        let value = tokens.[index].Value
        index <- index + 1
        value

    member this.CurrentKind =
        this.SkipNewlines()
        tokens.[index].Kind

    member this.ReadPropertyKey(?allowQuoted: bool) =
        let allowQuoted = defaultArg allowQuoted false
        this.SkipNewlines()
        if allowQuoted && tokens.[index].Kind = TokenKind.String then this.ReadString()
        else this.ReadIdent()

    member this.ReadScalarValue() =
        this.SkipNewlines()
        match tokens.[index].Kind with
        | TokenKind.String -> this.ReadString()
        | TokenKind.Ident -> this.ReadIdent()
        | TokenKind.RelativeDay ->
            let value = tokens.[index].Value
            index <- index + 1
            value
        | _ -> raise (this.Unexpected "scalar value")

    member this.ReadHexColor() =
        this.SkipNewlines()
        if tokens.[index].Kind <> TokenKind.HexColor then
            raise (this.Unexpected "#rrggbb or #rgb hex color")
        let value = tokens.[index].Value
        index <- index + 1
        value

    member this.ReadQualifiedName() =
        this.SkipNewlines()
        match tokens.[index].Kind with
        | TokenKind.Ident | TokenKind.Raw ->
            let value = tokens.[index].Value
            index <- index + 1
            value
        | _ -> raise (this.Unexpected "qualified name")

    member this.TryPeekIdent() =
        this.SkipNewlines()
        if tokens.[index].Kind <> TokenKind.Ident then None
        else Some tokens.[index].Value

    member this.ReadRawBlock() =
        this.SkipNewlines()
        if tokens.[index].Kind <> TokenKind.Raw then
            raise (this.Unexpected "[[ … ]] raw block")
        let value = tokens.[index].Value
        index <- index + 1
        value

    member private this.ReadRelativeDay() =
        this.SkipNewlines()
        if tokens.[index].Kind <> TokenKind.RelativeDay then
            raise (this.Unexpected "relative day")
        let value = tokens.[index].Value
        index <- index + 1
        value

    member private this.ReadDateBoundToken() =
        this.SkipNewlines()
        match tokens.[index].Kind with
        | TokenKind.RelativeDay -> this.ReadRelativeDay()
        | TokenKind.Ident -> this.ReadIdent()
        | _ -> raise (this.Unexpected "date bound (today, -Nd, yyyy-MM-dd)")

    member this.ReadDateDefaultValue() =
        let from = this.ReadDateBoundToken()
        if this.RawKind <> TokenKind.DotDot then from
        else
            index <- index + 1
            let toValue = this.ReadDateBoundToken()
            $"{from}..{toValue}"

    member private this.ReadListItem() =
        this.SkipNewlines()
        match tokens.[index].Kind with
        | TokenKind.Ident -> this.ReadIdent()
        | TokenKind.String -> this.ReadString()
        | _ -> raise (this.Unexpected "list item")

    member this.ReadCommaSeparatedValues() =
        let parts = ResizeArray<string>()
        parts.Add(this.ReadListItem())
        while this.CurrentKind = TokenKind.Comma do
            index <- index + 1
            parts.Add(this.ReadListItem())
        String.Join(", ", parts)

    member this.ReadCommaListInline() =
        let names = ResizeArray<string>()
        names.Add(this.ReadIdent())
        while this.CurrentKind = TokenKind.Comma do
            index <- index + 1
            names.Add(this.ReadIdent())
        names :> IReadOnlyList<_>

    member this.ReadColumnBinding() =
        let column = this.ReadQualifiedName()
        if this.TryKeyword "as" then
            { Column = column; Alias = Some(this.ReadString()) }
        elif tokens.[index].Kind = TokenKind.Ident && not (this.IsOnNewline()) then
            let saved = this.SavePosition()
            let alias = this.ReadIdent()
            if this.RawKind = TokenKind.Eq then
                this.RestorePosition saved
                { Column = column; Alias = None }
            elif String.Equals(alias, "end", StringComparison.OrdinalIgnoreCase) then
                this.RestorePosition saved
                { Column = column; Alias = None }
            else
                { Column = column; Alias = Some alias }
        else
            { Column = column; Alias = None }

    member this.ReadRestOfLine() =
        let parts = ResizeArray<string>()
        while tokens.[index].Kind = TokenKind.Ident || tokens.[index].Kind = TokenKind.Raw do
            parts.Add(tokens.[index].Value)
            index <- index + 1
        String.Join(' ', parts)

    member this.TryModuleInclude() =
        this.SkipNewlines()
        if this.TryKeyword "import" then
            Some(this.ReadString())
        elif not (this.IsAt TokenKind.Bang) then
            None
        else
            this.Advance()
            if not (this.TryKeyword "include") then
                raise (DashSpecParseException("Expected 'include' after '!'."))
            Some(this.ReadString())

    member this.SkipFileDirectives() =
        this.SkipNewlines()
        let mutable continueDirectives = true

        while continueDirectives && this.IsAt TokenKind.At do
            this.Advance()

            if this.TryKeyword "runtime" || this.TryKeyword "config" then
                if consumedRuntimePath.IsSome then
                    raise (DashSpecParseException("Only one @runtime directive is allowed per .dashspec file."))
                if this.IsAt TokenKind.String then
                    consumedRuntimePath <- Some(this.ReadString())
                this.SkipNewlines()
            elif this.TryKeyword "sqldialect" then
                this.ReadIdent() |> ignore
                this.SkipNewlines()
            elif this.TryKeyword "diagramlibrary" then
                if consumedDiagramLibraryPath.IsSome then
                    raise (DashSpecParseException("Only one @diagramlibrary directive is allowed per .dashspec file."))
                if this.IsAt TokenKind.String then
                    consumedDiagramLibraryPath <- Some(this.ReadString())
                this.SkipNewlines()
            elif this.TryKeyword "palette" then
                this.SkipNewlines()

                if this.RawKind = TokenKind.String then
                    if consumedPalettePath.IsSome then
                        raise (DashSpecParseException("Only one @palette file directive is allowed per .dashspec file."))
                    consumedPalettePath <- Some(this.ReadString())
                    this.SkipNewlines()
                else
                    index <- index - 2
                    continueDirectives <- false
            else
                index <- index - 1
                continueDirectives <- false